code/examples/misc_8.rs

```diff
@@ -0,0 +1,39 @@
+use core::pin::Pin;
+use core::marker::PhantomPinned;
+
+#[derive(Default)]
+struct MyPinnedType {
+    addr: usize,
+    _pin: PhantomPinned,
+}
+
+impl MyPinnedType {
+    fn my_pin_fn(self: Pin<&mut Self>) {
+        let me = unsafe { Pin::into_inner_unchecked(self) };
+        let me_addr = me as *mut Self as usize;
+        if me.addr == 0 {
+            me.addr = me_addr;
+        } else {
+            assert_eq!(me.addr, me_addr, "Pinned value was moved.")
+        }
+    }
+}
+
+trait MyUnpinTrait {
+    fn into_pinned_type(self: Pin<&mut Self>) -> Pin<&mut MyPinnedType>;
+}
+impl MyUnpinTrait for MyPinnedType {
+    fn into_pinned_type(self: Pin<&mut Self>) -> Pin<&mut MyPinnedType> {
+        self
+    }
+}
+impl Unpin for dyn MyUnpinTrait {}
```

Comment:
given that the compiler error is here, I think it should be simplified a bit to make it easier to understand